### DISCUSSION TITLE
Cannot install module because it is not a properly-formed module

### DISCUSSION THREAD

**Discussion by grzegorzwierzbicki** [pnp/powershell#453] [General]
Hi
I am trying to install the module on a Windows 2012 R2 server and run into this issue.

Install-Module pnp.powershell -Scope AllUsers -AcceptLicense -Verbose

VERBOSE: Acquiring providers for assembly: C:\Program Files\WindowsPowerShell\Modules\PackageManagement\fullclr\Microsoft.PackageManagement.ArchiverProviders.dll
VERBOSE: Acquiring providers for assembly: C:\Program Files\WindowsPowerShell\Modules\PackageManagement\fullclr\Microsoft.PackageManagement.MetaProvider.PowerShell.dll
VERBOSE: Acquiring providers for assembly: C:\Program Files\WindowsPowerShell\Modules\PackageManagement\fullclr\Microsoft.PackageManagement.NuGetProvider.dll
VERBOSE: Acquiring providers for assembly: C:\Program Files\WindowsPowerShell\Modules\PackageManagement\fullclr\Microsoft.PackageManagement.CoreProviders.dll
VERBOSE: Acquiring providers for assembly: C:\Program Files\WindowsPowerShell\Modules\PackageManagement\fullclr\Microsoft.PackageManagement.MsuProvider.dll
VERBOSE: Acquiring providers for assembly: C:\Program Files\WindowsPowerShell\Modules\PackageManagement\fullclr\Microsoft.PackageManagement.MsiProvider.dll
VERBOSE: Using the provider 'PowerShellGet' for searching packages.
VERBOSE: The -Repository parameter was not specified.  PowerShellGet will use all of the registered repositories.
VERBOSE: Getting the provider object for the PackageManagement Provider 'NuGet'.
VERBOSE: The specified Location is 'https://www.powershellgallery.com/api/v2' and PackageManagementProvider is 'NuGet'.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='pnp.powershell'' for ''.
VERBOSE: Total package yield:'1' for the specified package 'pnp.powershell'.
VERBOSE: Getting the provider object for the PackageManagement Provider 'NuGet'.
VERBOSE: The specified Location is 'https://www.poshtestgallery.com/api/v2/' and PackageManagementProvider is 'NuGet'.
VERBOSE: Searching repository 'https://www.poshtestgallery.com/api/v2/FindPackagesById()?id='pnp.powershell'' for ''.
VERBOSE: Total package yield:'0' for the specified package 'pnp.powershell'.
VERBOSE: Performing the operation "Install-Module" on target "Version '1.4.0' of module 'PnP.PowerShell'".
VERBOSE: The installation scope is specified to be 'AllUsers'.
VERBOSE: The specified module will be installed in 'C:\Program Files\WindowsPowerShell\Modules'.
VERBOSE: The specified Location is 'NuGet' and PackageManagementProvider is 'NuGet'.
VERBOSE: Downloading module 'PnP.PowerShell' with version '1.4.0' from the repository 'https://www.powershellgallery.com/api/v2'.
VERBOSE: Searching repository 'https://www.powershellgallery.com/api/v2/FindPackagesById()?id='PnP.PowerShell'' for ''.
VERBOSE: InstallPackage' - name='PnP.PowerShell', version='1.4.0',destination='C:\Users\xxxxxxxxx\AppData\Local\Temp\2053916661'
VERBOSE: DownloadPackage' - name='PnP.PowerShell',version='1.4.0',destination='C:\Users\xxxxxxxxx\AppData\Local\Temp\2053916661\PnP.PowerShell.1.4.0\PnP.PowerShell.1.4.0.nupkg',uri='https://www.powershellgallery.com/api/v2/package/PnP.PowerShell/1.4.0'
VERBOSE: Downloading 'https://www.powershellgallery.com/api/v2/package/PnP.PowerShell/1.4.0'.
VERBOSE: Completed downloading 'https://www.powershellgallery.com/api/v2/package/PnP.PowerShell/1.4.0'.
VERBOSE: Completed downloading 'PnP.PowerShell'.
VERBOSE: Hash for package 'PnP.PowerShell' does not match hash provided from the server.
VERBOSE: InstallPackageLocal' - name='PnP.PowerShell', version='1.4.0',destination='C:\Users\xxxxxxxxx\AppData\Local\Temp\2053916661'
VERBOSE: Validating the 'PnP.PowerShell' module contents under 'C:\Users\xxxxxxxxx\AppData\Local\Temp\2053916661\PnP.PowerShell.1.4.0' path.
PackageManagement\Install-Package : The module 'PnP.PowerShell' cannot be installed or updated because it is not a properly-formed module.
At C:\Program Files\WindowsPowerShell\Modules\PowerShellGet\PSModule.psm1:9711 char:34
    +             $installedPackages = PackageManagement\Install-Package @PSBoundParam ...
    +                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (Microsoft.Power....InstallPackage:InstallPackage) [Install-Package], Exception
    + FullyQualifiedErrorId : InvalidManifestModule,Test-ValidManifestModule,Microsoft.PowerShell.PackageManagement.Cmdlets.InstallPackage

**Comment by grzegorzwierzbicki** [comment]
Small update: the issue is NOT occurring when using PowerShell 7.0.6

**Comment by ToddKlindt** [comment]
This message makes me think it might be a transient error:
`VERBOSE: Hash for package 'PnP.PowerShell' does not match hash provided from the server.`
Can you reproduce it?

**Reply by grzegorzwierzbicki** [reply]
Yes, I can reproduce it.
In fact, the same thing is happening on a different WS 2012 R2 that I manage.

I thought it might be network or proxy related so I made sure my standard network tweaks are applied.
`
[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
[System.Net.WebRequest]::DefaultWebProxy = [System.Net.WebRequest]::GetSystemWebProxy()
[System.Net.WebRequest]::DefaultWebProxy.Credentials = [System.Net.CredentialCache]::DefaultNetworkCredentials`
That did not help.

I tried to force the older version with `-MaximumVersion 1.3.0`
No go.

I copied the PnP.PowerShell.1.4.0.nupkg file from temp location while the installer was running.
File is 21717358 bytes and sha256 for it is:
`E826C2A948CB00B077152980C27A30BF90E89EFBD7A0CED9A0B100AE7CF40069`
This hash is consistent between different servers.

**Reply by ToddKlindt** [reply]
It seems the hash error might be a red herring then. :(

**Reply by AndersRask** [reply]
If you check the part of the error:
`The module 'PnP.PowerShell' cannot be installed or updated because it is not a properly-formed module.`
It looks like you have an old module in your 5.1 module path that is malformed. Try and type
`Get-Module PnP.PowerShell -ListAvailable`
to see all versions you have installed, and remove old versions.
This would also explain why PS7.1 works as this has a different module path.

**Reply by grzegorzwierzbicki** [reply]
Nothing there that would conflict

dir "C:\Program Files\WindowsPowerShell\Modules"


`Directory: C:\Program Files\WindowsPowerShell\Modules
Mode                LastWriteTime     Length Name
----                -------------     ------ ----
d----        24/06/2020     13:37            7Zip4Powershell
d----        30/12/2020     13:32            AzureAD
d----        18/09/2020     16:07            AzureADPreview
d----        30/12/2020     13:22            CloudConnect
d----        30/03/2020     06:20            CredentialManager
d----        01/10/2020     10:01            ExchangeOnlineManagement
d----        26/03/2020     17:47            MSOnline
d----        26/03/2020     17:31            NuGet
d----        17/03/2021     13:15            PackageManagement
d----        17/03/2021     13:08            PowerShellGet`

**Reply by AndersRask** [reply]
There are more than one path to where they could be installed, depending on setup, so checking with cmdlet
`Get-Module PnP.PowerShell -ListAvailable`
Is usually better
`$env:PSModulePath `
shows all valid paths for the shell you have open.

**Reply by grzegorzwierzbicki** [reply]
`Get-Module PnP.PowerShell -ListAvailable`  returns nothing

**Reply by AndersRask** [reply]
Please check the temp folder mentioned in the error, and see if files are blocked, and if so unblock them:

`Get-ChildItem -Recurse | Unblock-File`

Another thing to check is your execution policy for the server at least allows RemoteSigned

**Reply by fs366e2spm** [reply]
I'm facing exactly the same warnings
`VERBOSE: Hash for package 'PnP.PowerShell' does not match hash provided from the server.
WARNING: The module 'PnP.PowerShell' cannot be installed or updated because it is not a properly-formed module.
`

PowerShell Version:
Major  Minor  Build  Revision
-----  -----  -----  --------
5       0          10586        117   

Get-Module PnP.PowerShell -ListAvailable returns nothing

**Reply by fs366e2spm** [reply]
@grzegorzwierzbicki did you find a way to install it?
